sling/nlp/parser/trainer/pytorch_modules.py

```diff
@@ -385,105 +471,75 @@ def forward(self, document, train=False):
-        state.advance(action)
-
-      return state
-
-
-  # Traces the model as it runs through 'document'.
-  def model_trace(self, document):
-    length = document.size()
-    lr_out, rl_out, lstm_features = self._lstm_outputs(document)
-
-    assert len(self.spec.lstm_features) == len(lstm_features)
-    for f in lstm_features:
-      assert len(f.offsets) == length
-
-    print length, "tokens in document"
-    for index, t in enumerate(document.tokens()):
-      print "Token", index, "=", t.text
-    print
-
-    state = ParserState(document, self.spec)
-    actions = self.spec.actions
-    ff_activations = []
-    steps = 0
-    for gold in document.gold:
-      print "State:", state
-      gold_index = actions.indices.get(gold, None)
-      assert gold_index is not None, "Unknown gold action %r" % gold
-
-      if state.current < state.end:
-        print "Token", state.current, "=", document.tokens()[state.current].text
-        for feature_spec, values in zip(self.spec.lstm_features, lstm_features):
-          # Recall that 'values' has indices at all sequence positions.
-          # We need to get the slice of feature indices at the current token.
-          start = values.offsets[state.current - state.begin]
-          end = None
-          if state.current < state.end - 1:
-            end = values.offsets[state.current - state.begin + 1]
-
-          current = values.indices[start:end]
-          print "  LSTM feature:", feature_spec.name, ", indices=", current,\
-              "=", self.spec.lstm_feature_strings(current)
-
-      ff_output, ff_debug = self._ff_output(
-          lr_out, rl_out, ff_activations, state, debug=True)
-      for f, indices in ff_debug:
-        debug = self.spec.ff_fixed_features_debug(f, indices)
-        print "  FF Feature", f.name, "=", str(indices), debug
-      assert ff_output.view(1, -1).size(1) == self.spec.num_actions
-
-      assert state.is_allowed(gold_index), "Disallowed gold action: %r" % gold
-      state.advance(gold)
-      print "Step", steps, ": advancing using gold action", gold
-      print
-      steps += 1
+      shift = actions.action(actions.shift())
+      stop = actions.action(actions.stop())
+      disallowed_counts = [0] * cascade.size()
+      total_counts = [0] * cascade.size()
+      while not state.done:
+        # Compute the hidden layer once for all cascade delegates.
+        ff_hidden, _ = self._ff_hidden(lr_out, rl_out, ff_activations, state)
+        cascading = True
+        delegate_index = 0
+
+        # Store the last CASCADE action in a cascade.
+        last = None
+        while cascading:
+          # Get the highest scoring action from the cascade delegate.
+          # Note: We don't have to do any filtering or checking here, we
+          # can just return the top-scoring action.
+          best = cascade.predict(delegate_index, state, last, ff_hidden)
+
+          if best.is_cascade():
+            delegate_index = cascade.next(best, delegate_index)
```

Comment:
Why isn't the delegate index not just encoded in the action? Here you instead call the cascade to determine the next delegate index.

Reply:
Done.